Fixing the debugger "wrong namespace" issues.

vspinu opened the pull request:
Fixing #420,  clojure-emacs/cider#2039 and clojure-emacs/cider#2004

expez commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -189,7 +190,7 @@ this map (identified by a key), and will `dissoc` it afterwards."}
   type is sent in the message as :input-type."
   [extras type prompt]
-  (let [key (u/random-uuid-str)
+  (let [key   (u/random-uuid-str)
```
Think you forgot to turn off your alignment minor mode :)

bbatsov replied:
It's not really a minor mode - just a toggle in `clojure-mode`. ;-)

vspinu replied:
What's wrong with this? You don't want aligned key-val pairs in cider projects?

bbatsov replied:
Yes. I'm not fond of this style and avoid it everywhere.

expez commented:
Think a note in the changelog for the next cider release would be nice too.  I know a lot of people have been tripped up by this one.

Great stuff as usual, @vspinu! :+1:

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -227,18 +237,19 @@ this map (identified by a key), and will `dissoc` it afterwards."}
       (finally
         (try
-          (in-ns ns)
+          ;; Restore original ns in case it was changed by the evaled code.
```
evaled -> evaluated

vspinu replied:
Was saving few chars not to get into fill.

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -412,19 +426,22 @@ this map (identified by a key), and will `dissoc` it afterwards."}
-                  :forms @*tmp-forms*}]
+  `(let [~'META__ {:msg ~(let [{:keys [code id file line column ns]} *msg*]
+                           (-> {:code code,
+                                ;; Passing namespace as :original-ns stalls. Bug?
```
stalls what?

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -583,12 +602,7 @@ this map (identified by a key), and will `dissoc` it afterwards."}
                  (h (maybe-debug msg)))
       "debug-instrumented-defs" (instrumented-defs-reply msg)
       "debug-input" (when-let [pro (@promises (:key msg))]
-                      (swap! promises dissoc  (:key msg))
```
How do those changes fit with the rest?

vspinu replied:
There is no rest, the only consumer of that promise is the `read-debug` itself.

bbatsov replied:
I see.

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -536,12 +536,14 @@ this map (identified by a key), and will `dissoc` it afterwards."}
+  (let [has-debug?  (atom false)
+        ;; Don't return nil in reader (https://dev.clojure.org/jira/browse/CLJ-1138)
+        fake-reader (fn [x] (reset! has-debug? true) x)]
+    (binding [*ns*           (find-ns (symbol (or ns "user")))
```
We don't really use alignment in bindings in `cider-nrepl`.

vspinu replied:
Let bindings as well?

bbatsov replied:
Yes.

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -207,12 +207,14 @@ this map (identified by a key), and will `dissoc` it afterwards."}
-      (binding [*tmp-locals* locals]
+      (binding [*tmp-locals* (:locals extras)
+                ;; evaluate in the instrumentation ns
+                *ns*         (find-ns (symbol (:original-ns extras)))]
```
Same here. What will happen if `:original-ns` is nil here?

vspinu replied:
`:original-ns` can never be nil, it's always assigned in `with-initial-debug-bindings`.

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -207,12 +207,14 @@ this map (identified by a key), and will `dissoc` it afterwards."}
-  [form locals]
+  If an exception is thrown, it is caught and sent to the client, and this
+  function returns nil. `extras` is a metadata map received from `break'."
+  [form extras]
```
I'd name `extras` something else - without reading its docstring it was hard for me to figure out what it is. `break-state` perhaps?

vspinu replied:
Good sugestion. That map contains both current breakpoint state and current debug session state, but I think `break-state` is informative of both.

bbatsov replied:
Maybe `debugger-state` then with two maps inside of it?

bbatsov commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -227,18 +229,19 @@ this map (identified by a key), and will `dissoc` it afterwards."}
 (defn- read-debug-eval-expression
   "Read and eval an expression from the client.
-  extras is a map to be added to the message, and prompt is added into
+  `extras` is a map received from `break`, and `prompt` is added into
```
Same here.

bbatsov commented:
> Think a note in the changelog for the next cider release would be nice too. I know a lot of people have been tripped up by this one.

Yeah, once this is merged it should have a changelog entry for sure. Probably it will be the final bug-fix for 0.15.1. Seems no other major problems with 0.15 have been discovered in the weeks after its release.

vspinu commented:
Good to go AFAIC. Addressed comments, fixed tests and added some extras:
 - improved and disentangled error handling during debug eval commands (eval, inspect, inject, locals), 
 - refactored naming conventions here and there

Malabarba commented on src/cider/nrepl/middleware/debug.clj:
```diff
@@ -185,20 +183,41 @@ this map (identified by a key), and will `dissoc` it afterwards."}
+                          {:status :eval-error
+                           :causes [(let [causes# (stacktrace/analyze-causes e# (:pprint-fn *msg*))]
+                                      (when (coll? causes#) (last causes#)))]})))
+                    ::error))]
```
It would be more robust (and just as easy) to use a generated symbol here.

vspinu replied:
Good point. Changed.

Malabarba commented:
LGTM

bbatsov commented:
👍
